### PR TITLE
support yaml definition of Outputs with Deep expression property

### PR DESCRIPTION
**Pull request by dwy12321**
**Describe the change**
Add Outputs assign with deep expression property.
As yaml definition workflow , i can not assign data to  custom data's sub sub property relevant to  workflow instance.
As following example,  when i assgin 0 to WorkflowCore.Sample03.MyDataClass sub sub property Value6, i get an exception.

DataType: WorkflowCore.Sample03.MyDataClass, WorkflowCore.Sample03
Steps:
-- Id: HelloWorldId
  Name: HelloWorld
  StepType: WorkflowCore.Sample03.Steps.HelloWorld, WorkflowCore.Sample03
  Outputs: 
    **myData.Value6: 0**

**Describe your implementation or design**
How did you go about implementing the change?

**Tests**
Did you cover your changes with tests?

**Breaking change**
 compatibility with previous versions?

### CODE REVIEW

**Review by danielgerlag — COMMENTED** [review]
Could we add some tests to demonstrate the new functionality working and also make sure it's not breaking indexed properties?  Could we also clean up the formatting?

**Comment by dwy12321** [comment]
Hi  daniel,
I add WorkflowCore.Tests.YmalDefinition to test this change, In this test case, the data transmission is correct. Could you please help review this pr.

File: test/WorkflowCore.Tests.YmalDefinition/YmalDefinitionTest.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using WorkflowCore.Interface;
+using WorkflowCore.Testing;
+using Xunit;
+
+namespace WorkflowCore.Tests.YmalDefinition
+{
+    public class YmalDefinitionTest:YamlWorkflowTest
```
Comment (danielgerlag) [inline test/WorkflowCore.Tests.YmalDefinition/YmalDefinitionTest.cs:14]: The changes you made are not YAML specific, they apply to the definition loader, regardless of serialization format.  I think we should add unit tests [https://github.com/danielgerlag/workflow-core/blob/master/test/WorkflowCore.UnitTests/Services/DefinitionStorage/DefinitionLoaderTests.cs](here). This particular test is more of an integration test, which we could add to the integration tests project.  I think the unit tests are more important in this case.

Comment (dwy12321) [reply test/WorkflowCore.Tests.YmalDefinition/YmalDefinitionTest.cs]: Thanks,good suggesttion, I have  added to  unit tests.

File: src/WorkflowCore.DSL/Services/DefinitionLoader.cs [hunk 2]
```diff
@@ -217,17 +217,76 @@ private void AttachOutputs(StepSourceV1 source, Type dataType, Type stepType, Wo
...
                 // Check if our datatype has a matching property
-                var propertyInfo = dataType.GetProperty(output.Key);
-                if (propertyInfo != null)
+
+                PropertyInfo propertyInfo = null;
+                String[] paths = output.Key.Split('.');
+
+                bool hasAddOutput = false;
+
+                foreach (String propertyName in paths)
```
Comment (danielgerlag) [inline src/WorkflowCore.DSL/Services/DefinitionLoader.cs:229]: This method is getting quite big and doing lots of things... could we split the nested property flattening into it's own method?

File: test/WorkflowCore.UnitTests/Services/DefinitionStorage/DefinitionLoaderTests.cs [hunk 3]
```diff
@@ -43,6 +43,16 @@ public void ParseDefinition()
             A.CallTo(() => _registry.RegisterWorkflow(A<WorkflowDefinition>.That.Matches(MatchTestDefinition, ""))).MustHaveHappened();
         }
 
+        [Fact(DisplayName = "Should parse definition")]
+        public void ParseDefinitionPropertyDynamic()
```
Comment (danielgerlag) [inline test/WorkflowCore.UnitTests/Services/DefinitionStorage/DefinitionLoaderTests.cs:47]: Should we also test that the nested output properties are getting mapped according to your change?

Comment (dwy12321) [reply test/WorkflowCore.UnitTests/Services/DefinitionStorage/DefinitionLoaderTests.cs:47]: Ok , I add nested output properties in  test

File: src/WorkflowCore.DSL/Services/DefinitionLoader.cs (hunk 2 above)
Comment (dwy12321) [reply src/WorkflowCore.DSL/Services/DefinitionLoader.cs]: ok, split the nested property and direct property using AttachNestedOutput AttachDirectlyOutput

**Comment by dwy12321** [comment]
Hi daniel,
Would you please help review this pull request?

**Pull request merged**